### PR TITLE
Closes #25809: tapping on search engine triggers engine selection

### LINKED ISSUE #25809: Tapping on search engine icon should not enter input mode
When user tap on the search engine icon in the toolbar, we should allow search engine selection menu popup without entering input mode. (No virtual keyboard should show)

┆Issue is synchronized with this [Jira Task](https://mozilla-hub.atlassian.net/browse/FNXV2-20847)

### PR DESCRIPTION
**Pull request by mavduevskiy**
Adding search selector to the toolbar in `Display mode`. Flow is well described [here](https://www.figma.com/file/epu0Sj6mCPk7K282sNpHgI/Site?node-id=3851%3A77540#296011078).

Old|New
---|---
unified_on_old|unified_on_new
unified_off_old|unified_off_new

There is also a spacing issue that is particularly visible on devices that have arrow down keyboard button.
https://user-images.githubusercontent.com/92760693/199746717-05e0c70e-ce85-4056-830f-e8342fb3e0e6.mp4

I could make it consistent by adding more spacing to the display mode , but not sure if I should. An easy fix would look like this. 
https://user-images.githubusercontent.com/92760693/199754394-c09b423e-b36e-4efe-812d-065be20850d0.mp4

Will take a look if it's possible to fix the padding in the edit mode (so that an icon without selector wouldn't take extra space), as a follow up.

### Pull Request checklist

- [ ] **Tests**: This PR includes thorough tests or an explanation of why it does not
- [ ] **Screenshots**: This PR includes screenshots or GIFs of the changes made or an explanation of why it does not
- [ ] **Accessibility**: The code in this PR follows [accessibility best practices](https://github.com/mozilla-mobile/shared-docs/blob/master/android/accessibility_guide.md) or does not include any user facing features. In addition, it includes a screenshot of a successful [accessibility scan](https://play.google.com/store/apps/details?id=com.google.android.apps.accessibility.auditor&hl=en_US) to ensure no new defects are added to the product.

### QA

- [x] **QA Needed**

### To download an APK when reviewing a PR (after all CI tasks finished running):
1. Click on `Checks` at the top of the PR page.
2. Click on the `firefoxci-taskcluster` group on the left to expand all tasks.
3. Click on the `build-debug` task.
4. Click on `View task in Taskcluster` in the new `DETAILS` section.
5. The APK links should be on the right side of the screen, named for each CPU architecture.

### GitHub Automation
Fixes #25809

### CODE REVIEW

File: app/src/main/java/org/mozilla/fenix/search/SearchFragmentStore.kt
Comment (gabrielluong) [inline app/src/main/java/org/mozilla/fenix/search/SearchFragmentStore.kt:126]:
```suggestion
    searchEngine: SearchEngine?,
```
Could we also reorder this to be below `searchAccessPoint` to align with the param ordering you did above in `SearchDialogFragment`

File: app/src/main/java/org/mozilla/fenix/search/toolbar/SearchSelectorMenu.kt
Comment (gabrielluong) [inline app/src/main/java/org/mozilla/fenix/search/toolbar/SearchSelectorMenu.kt:73]: Let's move this to its own file

File: app/src/main/java/org/mozilla/fenix/home/HomeFragment.kt
Comment (gabrielluong) [inline app/src/main/java/org/mozilla/fenix/home/HomeFragment.kt:370]: I have an interest to see more code refactored out of the `HomeFragment` and into their corresponding classes so that we could encapsulate all the logic around the search selector into the appropriate classes. Would this be possible here? Would it make sense add a public update function of some sort in `SearchSelectorMenu` or something else that can be called here?

Comment (gabrielluong) [reply app/src/main/java/org/mozilla/fenix/home/HomeFragment.kt:370]: Actually, this might fit more with the `Builder` classes we did for the `HomeFragment`. See `TabCounterBuilder` and `HomeMenuBuilder`.

Comment (gabrielluong) [reply app/src/main/java/org/mozilla/fenix/home/HomeFragment.kt:370]: We don't necessarily have to do this in this PR. This is more food for thought if we were able to encapsulate the logic here better so that our enormous `HomeFragment` won't hold so much of the logic related to various functions.

File: app/src/main/java/org/mozilla/fenix/search/toolbar/SearchSelectorMenu.kt
Comment (mavduevskiy) [reply app/src/main/java/org/mozilla/fenix/search/toolbar/SearchSelectorMenu.kt:75]: Could you explain a bit more? I see a lot of classes (that relate to the UI, like views and delegates) that take interactor as a parameter declaring the interface inside the file. Well, even this one was chipped away from `ToolbarInteractor`, which is the same file with `ToolbarView`.

File: app/src/main/java/org/mozilla/fenix/home/HomeFragment.kt
Comment (mavduevskiy) [reply app/src/main/java/org/mozilla/fenix/home/HomeFragment.kt:370]: Yes, I also think that it has to be encapsulated. Actually, it is – inside `SearchSelectorToolbarAction`, which is a `Toolbar.Action` and can't be used on a "fake" toolbar we have on the homescreen. Creating a more real toolbar for the `HomeFragment` could be a direction to go? I have little context why it was implemented that way.

File: app/src/main/java/org/mozilla/fenix/search/toolbar/SearchSelectorMenu.kt
Comment (gabrielluong) [reply app/src/main/java/org/mozilla/fenix/search/toolbar/SearchSelectorMenu.kt:75]: This is more about moving the Interactor interface into its own file and keeping them out of the views. We have found that there will be times where the Interactor interface could be reused in a variety of places so it would make more sense to keep them to their own file from the get go.

I would imagine you are referring to Bookmark and History views for what you're seeing, but those are also some of our oldest views that we don't actively polish compare to the Home screen. Other than that I would say we would rather keep files smaller for better development ergonomics if there's a combination of classes, functions and interfaces that are all in one place. I could also imagine the views are more volatile to changes and deletion compare to its own separate file that holds the interactor interface.

Ultimately, this is just about developer ergonomics. It's easier to work with smaller files and you aren't at risk of refactoring in the future.

File: app/src/main/java/org/mozilla/fenix/home/HomeFragment.kt
Comment (gabrielluong) [reply app/src/main/java/org/mozilla/fenix/home/HomeFragment.kt:370]: When I wrote `SearchSelectorToolbarAction`, it was for encapsulating the `Toolbar.Action` functionality that would be added to the edit toolbar actions. This is indeed a direction I am trying to get at, but I am getting the sense now that we should probably be a bit less concern about the ideas I presented in this conversation for the sake of this PR and we can revisit this more in a future cleanup of `HomeFragment`. We'll want to think about what makes sense to group up and where (should it be be done based on the component in the `binding`, etc). 

We can resolve this conversation.

File: app/src/main/java/org/mozilla/fenix/home/HomeFragment.kt
Comment (gabrielluong) [inline app/src/main/java/org/mozilla/fenix/home/HomeFragment.kt:673]: I think we should stick to using a simple if statement check for `searchEngine != null` here instead of the `let` follow by the `run`

**Comment by mavduevskiy** [comment]
At the same time, [setBounds](https://searchfox.org/mozilla-mobile/source/fenix/app/src/main/java/org/mozilla/fenix/home/HomeFragment.kt#615) method does not have any effect on the `BitmapDrawable` size. It keeps the Bitmap the initial size, and `BitmapDrawable` takes its parameters. 

We don't notice it with the icon in the `HomeFragment` because of the default scaling done by `ImageView`. But it has a visible effect on the search selector, as it uses specified scaling: `android:scaleType="center"`. It creates extra padding when is fitting a bigger icon into a smaller frame. 

The [extension](https://searchfox.org/mozilla-mobile/source/fenix/app/src/main/java/org/mozilla/fenix/search/toolbar/SearchSelectorToolbarAction.kt#90) actually works, because it creates a smaller Bitmap. But that disables the [test](https://searchfox.org/mozilla-mobile/source/fenix/app/src/androidTest/java/org/mozilla/fenix/ui/SettingsSearchTest.kt#74) that verifies bitmap in the `ImageView` matches the on of the search engine. 

I am removing not working `setBounds` code from the `HomeFragment` and scaling from the `search_selector`. It helps to keep the test and looks safe.

As a side note, why would  `search_selector` has a special scaling if it receives already scaled  by the extension `BitmapDrawable`? I might be missing something.

File: app/src/main/java/org/mozilla/fenix/search/toolbar/SearchSelectorMenu.kt
Comment (gabrielluong) [inline app/src/main/java/org/mozilla/fenix/search/toolbar/SearchSelectorMenu.kt:35]: nit no new line necessary

File: app/src/main/java/org/mozilla/fenix/home/HomeFragment.kt
Comment (MozillaNoah) [inline app/src/main/java/org/mozilla/fenix/home/HomeFragment.kt:362]:
```suggestion
                UnifiedSearch.searchMenuTapped.record(NoExtras())
```

Can we clean up the full import here?

File: app/src/main/java/org/mozilla/fenix/home/HomeFragment.kt
Comment (MozillaNoah) [inline app/src/main/java/org/mozilla/fenix/home/HomeFragment.kt:368]: should this just live in the XML?

File: app/src/main/java/org/mozilla/fenix/home/sessioncontrol/SessionControlController.kt
Comment (MozillaNoah) [inline app/src/main/java/org/mozilla/fenix/home/sessioncontrol/SessionControlController.kt:680]:
```suggestion
            SearchSelectorMenu.Item.SearchEngine -> {
```

**Review by MozillaNoah — COMMENTED** [review]
🚗 driving by with a couple comments

File: app/src/main/java/org/mozilla/fenix/home/sessioncontrol/SessionControlController.kt
Comment (mavduevskiy) [reply app/src/main/java/org/mozilla/fenix/home/sessioncontrol/SessionControlController.kt:680]: It does look odd, but `SearchEngine` is a data class, so won't work without `is`

**Comment by mavduevskiy** [comment]
@MozillaNoah @gabrielluong 
Thanks for taking your time to review! updated

File: app/src/main/java/org/mozilla/fenix/search/toolbar/SearchSelector.kt
Comment (gabrielluong) [inline app/src/main/java/org/mozilla/fenix/search/toolbar/SearchSelector.kt:25]: I am curious what kind of issues we ran into that required us to programmatically set the margin in the class. 
Could we also set the margin in the XML in `fragment_home` to avoid adding the code here?

File: app/src/main/res/layout/fragment_home.xml
Comment (gabrielluong) [inline app/src/main/res/layout/fragment_home.xml:148]: Could we set the marginTop here?

File: app/src/main/res/layout/fragment_home.xml
Comment (gabrielluong) [inline app/src/main/res/layout/fragment_home.xml:146]: Could we undo the spacing change here?

File: app/src/main/res/layout/search_selector.xml
Comment (gabrielluong) [inline app/src/main/res/layout/search_selector.xml:33]: Trying to understand this change. Could you explain what kind of image scale issues you were seeing here and what was the mitigation?

File: app/src/main/java/org/mozilla/fenix/home/HomeFragment.kt
Comment (gabrielluong) [inline app/src/main/java/org/mozilla/fenix/home/HomeFragment.kt:616]: Trying to understand if this change was intentional. I see we aren't setting the bounds anymore, but I am wondering what is the reasoning behind the change? I see we do call this type `setBounds(0, 0, iconSize, iconSize)` in many places in our codebase https://searchfox.org/mozilla-mobile/search?q=setBounds&path=, but I don't have any context on its significance or usage. Since we are removing it here, I would like to learn a bit more what types of problem we were encountering and solving.

File: app/src/main/java/org/mozilla/fenix/search/toolbar/SearchSelector.kt
Comment (mavduevskiy) [reply app/src/main/java/org/mozilla/fenix/search/toolbar/SearchSelector.kt:25]: `SearchSelector` has a hardcoded top margin, that made it placing in alongside the [searchIcon](https://searchfox.org/mozilla-mobile/source/fenix/app/src/main/res/layout/fragment_home.xml#138) difficult.

File: app/src/main/res/layout/fragment_home.xml
Comment (mavduevskiy) [reply app/src/main/res/layout/fragment_home.xml:148]: It won't affect the top margin of the material card inside the `[search_selector](https://searchfox.org/mozilla-mobile/source/fenix/app/src/main/res/layout/search_selector.xml#15)`

File: app/src/main/res/layout/search_selector.xml
Comment (mavduevskiy) [reply app/src/main/res/layout/search_selector.xml:33]: I tried to explain it [here](https://github.com/mozilla-mobile/fenix/pull/27647#issuecomment-1301656042).

In short, this scaling type would create extra margin for icons bigger then the view size. It can be seen with duck duck go icon, if one revers the change. Default scaling already works correctly for [searchIcon](https://searchfox.org/mozilla-mobile/source/fenix/app/src/main/res/layout/fragment_home.xml#138), so I took that approach

File: app/src/main/java/org/mozilla/fenix/home/HomeFragment.kt
Comment (mavduevskiy) [reply app/src/main/java/org/mozilla/fenix/home/HomeFragment.kt:616]: yes, I was talking about it [here](https://github.com/mozilla-mobile/fenix/pull/27647#issuecomment-1301656042).

short version, this particular code doesn't do anything, since we create a `BitmapDrawable` from a `Bitmap`. Changing it's bounds afterwards doesn't affect the Bitmap itself, so it's height and width will be used.
You can see that by changing `preference_icon_drawable_size` to 2 or 200 – it won't have any effect.

File: app/src/main/java/org/mozilla/fenix/search/toolbar/SearchSelector.kt
Comment (gabrielluong) [reply app/src/main/java/org/mozilla/fenix/search/toolbar/SearchSelector.kt:25]: This was a random lunch thought, but I haven't really looked into the code again with consideration of your comment yet. 

I am thinking in an ideal situation all of us would probably prefer to see all the layout properties defined in the XML layout. If another engineer was working on the search selector layout, they might be confused by where the top margin is coming from in their testing because they don't see the margin top 14dp defined in the XML. They would have to dig much deeper into the code in order to figure out that the margin is set manually in the code and that could be both surprising and maybe somewhat frustrating when working with this code. 

So, I am wondering if we could do some more exploring to see if we could define this margin in the XML. A naive idea I have would be to try `marginTop=-14dp` and see if that works? I don't necessarily have the answers right now, but I will take some time to play around with your PR to see if there is a reasonable solution when I have a chance. 

I am also happy to accept this as a followup investigation and it is possible that the only solution is to set the margin programatically. I just want to make sure we put some more thought into this since this will only improve how people work with this code.

File: app/src/main/res/layout/search_selector.xml
Comment (mavduevskiy) [reply app/src/main/res/layout/search_selector.xml:33]: here is the [patch](https://github.com/mozilla-mobile/fenix/pull/27726) with reverted scaling code. with unified search is on, duckduck, amazon, google icons will show the problem! one could also play around with `preference_icon_drawable_size` parameter to see if it has any effect.

**Comment by gabrielluong** [comment]
I haven't gotten around to the review because I am still looking to see if you had any response to 2 questions:
(1) Could we somehow set the margin setting in the XML instead of in the code? This is mostly to avoid surprises in terms of developer ergonomics when one is looking at the search selector layout properties in the XML and only learns that the margin top is set programmatic much later. https://github.com/mozilla-mobile/fenix/pull/27647#discussion_r1014374484
(2) I haven't had a chance to test and verify that `setBound` does nothing in relation to https://github.com/mozilla-mobile/fenix/pull/27647#discussion_r1013584431. I am mostly just suspicious of this change because this is the only line of code that will ship live because it is not behind a feature flag. It appears to be used commonly elsewhere as I have pointed out and it was added by pocmo by in the day so I have to imagine it was done for some purpose even though I do not yet understand why. So, my main ask here is to test this line of code a bit more and ensure we're still getting the expected results even in older Android versions. This means testing in Android 6 as stated in https://github.com/mozilla-mobile/fenix/pull/27726#issuecomment-1305120483.

**Comment by mavduevskiy** [comment]
@gabrielluong 
I created two follow ups, not to block this PR. 
1. [Investigate](https://github.com/mozilla-mobile/fenix/issues/27763) the top margin setting through xml.
2. [Investigate](https://github.com/mozilla-mobile/fenix/issues/27764) (apparently) not working scaling code in the HomeFragment.
 
Btw, even when approved, I don't think we should land it now. Next week merging into Beta looks like a safer option.

**Comment by gabrielluong** [comment]
> @gabrielluong I created two follow ups, not to block this PR. Reverted `setBounds` code, refactored.
> 
> 1. [Investigate](https://github.com/mozilla-mobile/fenix/issues/27763) the top margin setting through xml.
> 2. [Investigate](https://github.com/mozilla-mobile/fenix/issues/27764) (apparently) not working scaling code in the HomeFragment.
> 
> Btw, even when approved, I don't think we should land it now. Next week merging into Beta looks like a safer option.

Okay, I see that we reverted the `setBounds` code and will investigate it as a followup and that we will land this next week. I don't have concerns here. I will leave it to Roger to test and evaluate. Thank you for following up on the changes and investigation.

**Comment by mavduevskiy** [comment]
@gabrielluong 
Thank you for being so diligent. Looking back, the idea of removing code that looks like it is not doing anything as an act of random refactoring that is not really contributing to the task, does not look like a good one. It might not do anything on my device/OS/manufacturer, but it doesn't mean it won't break anything if we remove it. 

If I had good understanding why it was written, why we don't need it now and if it's safe to remove – that's another story. But we need more investigation, hence the follow up task. Lesson learnt👍

**Review by rocketsroger — APPROVED** [review]
Looks good to me, but let's land this after the cut.  Thanks

**Pull request merged**
